pandas/core/groupby.py

```diff
@@ -3740,9 +3740,31 @@ def _reindex_output(self, result):
             return result
 
         levels_list = [ping.group_index for ping in groupings]
-        index = MultiIndex.from_product(levels_list, names=self.grouper.names)
-        d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+        index, _ = MultiIndex.from_product(
+            levels_list, names=self.grouper.names).sortlevel()
+
+        if self.as_index:
+            d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
+            return result.reindex(**d)
+
+        # GH 13204
```

Comment:
```
In [7]: df = DataFrame({'cat': [1,2,3],'A': [10, 11, 11],'B': [101, 102, 103]})

In [8]: df.groupby(['cat','A']).sum()
Out[8]: 
          B
cat A      
1   10  101
2   11  102
3   11  103

In [9]: df.groupby(['cat','A'],as_index=False).sum()
Out[9]: 
   cat   A    B
0    1  10  101
1    2  11  102
2    3  11  103

In [10]: f = lambda r: df.loc[r, 'A'] % 2

In [11]: df.groupby(['cat',f],as_index=False).sum()
Out[11]: 
   cat   A    B
0    1  10  101
1    2  11  102
2    3  11  103
```

It appears non-named groupers (e.g. not ``in_axis``) are excluded, for non-categoricals (which makes sense its en emphereal grouper, not something we started with).